### PR TITLE
feat(stream): add debug logs for join key with high amplification

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Adapted from @yuhao-su's commit 4132beeaad6d42cce606807bfe2abaa20d239ed7, so we can have a general mechanism to debug this case.

It can be enabled with `RUST_LOG="hash_join_amplification=debug"`

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/executor/hash_join.rs [hunk 1]
```diff
@@ -829,6 +829,17 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
 
             if let Some(rows) = &matched_rows {
                 join_matched_join_keys.observe(rows.len() as _);
+                if rows.len() > 10000 {
+                    tracing::debug!(target: "hash_join_amplification",
+                        matched_rows_len = rows.len(),
+                        update_table_id = side_update.ht.table_id(),
+                        match_table_id = side_match.ht.table_id(),
+                        join_key = ?key,
```
Comment (yuhao-su) [inline src/stream/src/executor/hash_join.rs:837]: We can actually get a decoded value by using `row` and `join_key_indices`.

File: src/stream/src/executor/hash_join.rs [hunk 2]
```diff
@@ -829,6 +829,17 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
 
             if let Some(rows) = &matched_rows {
                 join_matched_join_keys.observe(rows.len() as _);
+                if rows.len() > 10000 {
```
Comment (yuhao-su) [inline src/stream/src/executor/hash_join.rs:832]: suggest setting to 65536

File: src/stream/src/executor/hash_join.rs (hunk 1 above)
Comment (kwannoel) [reply src/stream/src/executor/hash_join.rs:837]: Cool, let me have a try later

File: src/stream/src/executor/hash_join.rs (hunk 2 above)
Comment (kwannoel) [reply src/stream/src/executor/hash_join.rs:832]: 10K already seems quite large. It can enable us to capture a wider range of values.

Comment (kwannoel) [reply src/stream/src/executor/hash_join.rs:832]: No need to follow the boundaries of the bucket IMO

Comment (yuhao-su) [reply src/stream/src/executor/hash_join.rs:832]: Oh you are right. I thought it's 100k.

**Review by stdrc — APPROVED** [review]
LGTM

**Comment by lmatz** [comment]
do you want to include it into `v1.9.0`, we still got some time before releasing it today

File: src/stream/src/executor/hash_join.rs (hunk 1 above)
Comment (kwannoel) [reply src/stream/src/executor/hash_join.rs:837]: Follow up in future PR

**Comment by kwannoel** [comment]
> do you want to include it into `v1.9.0`, we still got some time before releasing it today

Hmm okay we can merge it in first.

File: src/stream/src/executor/hash_join.rs (hunk 1 above)
Comment (kwannoel) [reply src/stream/src/executor/hash_join.rs:839]: Updated https://github.com/risingwavelabs/risingwave/pull/16799/commits/977d15df8eae09f827df0619124a357314ce2940

**Pull request merged**